File: app/src/main/java/com/example/kolomachinetest/api/ApiCallback.java
package com.example.kolomachinetest.api;

import androidx.annotation.NonNull;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class ApiCallback<T> implements Callback<T> {

    private final RestCallback<T> mCallback;

    public ApiCallback(RestCallback<T> callback) {
        mCallback = callback;
    }


    public interface RestCallback<T> {
        void onSuccess(T result);

        void onFailure(String message);
    }

    @Override
    public void onResponse(@NonNull Call<T> call, @NonNull Response<T> response) {
        if (response.isSuccessful()) {
            mCallback.onSuccess(response.body());
        } else {
            mCallback.onFailure(response.message());
        }
    }

    @Override
    public void onFailure(@NonNull Call<T> call, @NonNull Throwable t) {
        mCallback.onFailure(t.getMessage());
    }
}

File: app/src/main/java/com/example/kolomachinetest/api/ApiCallback.kt
package com.example.kolomachinetest.api

import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class ApiCallback<T>(private val mCallback: CallBack<T>) : Callback<T> {

    override fun onResponse(call: Call<T>, response: Response<T>) {
        if (response.isSuccessful) {
            val result = response.body()
            if (result != null) {
                mCallback.onSuccess(result)
            }
        } else {
            mCallback.onFailure(response.message())
        }
    }

    override fun onFailure(call: Call<T>, t: Throwable) {
        mCallback.onFailure(t.message)
    }
}